controller/src/util/instance_action.rs

```diff
@@ -71,68 +72,68 @@ async fn internal_handle_existing_instances(
     // while.
-    while let Some(event) = stream.try_next().await? {
+    while let Some(event) = informer.try_next().await? {
         // Aquire lock to ensure cleanup_instance_and_configuration_svcs and the
         // inner loop handle_instance call in internal_do_instance_watch
         // cannot execute at the same time.
         let _lock = synchronization.lock().await;
         trace!("internal_do_instance_watch - aquired sync lock");
-        handle_instance(event, kube_interface).await?;
+        handle_instance(event, kube_interface, &mut first_event).await?;
     }
     Ok(())
 }
 
 /// This takes an event off the Instance stream and delegates it to the
 /// correct function based on the event type.
 async fn handle_instance(
-    event: WatchEvent<Instance>,
+    event: Event<Instance>,
     kube_interface: &impl KubeInterface,
-) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    first_event: &mut bool,
+) -> anyhow::Result<()> {
     trace!("handle_instance - enter");
     match event {
-        WatchEvent::Added(instance) => {
+        Event::Applied(instance) => {
             info!(
-                "handle_instance - added Akri Instance {:?}: {:?}",
+                "handle_instance - added or modified Akri Instance {:?}: {:?}",
                 instance.metadata.name, instance.spec
             );
+            // TODO: consider renaming `InstanceAction::Add` to `InstanceAction::AddOrUpdate`
+            // to reflect that this could also be an Update event. Or as we do more specific
+            // inspection in future, delineation may be useful.
             handle_instance_change(&instance, &InstanceAction::Add, kube_interface).await?;
-            Ok(())
         }
-        WatchEvent::Deleted(instance) => {
+        Event::Deleted(instance) => {
             info!(
                 "handle_instance - deleted Akri Instance {:?}: {:?}",
                 instance.metadata.name, instance.spec
             );
             handle_instance_change(&instance, &InstanceAction::Remove, kube_interface).await?;
-            Ok(())
         }
-        WatchEvent::Modified(instance) => {
-            info!(
-                "handle_instance - modified Akri Instance {:?}: {:?}",
-                instance.metadata.name, instance.spec
-            );
-            handle_instance_change(&instance, &InstanceAction::Update, kube_interface).await?;
-            Ok(())
-        }
-        WatchEvent::Error(ref e) => {
-            trace!("handle_instance - error for Akri Instance: {}", e);
-            Ok(())
+        Event::Restarted(_instances) => {
+            if *first_event {
+                info!("handle_instance - watcher started");
+            } else {
+                return Err(anyhow::anyhow!(
+                    "Instance watcher restarted - throwing error"
+                ));
+            }
         }
-        WatchEvent::Bookmark(_) => Ok(()),
     }
+    if *first_event {
```

Comment:
nit: not sure what is more expensive, branching logic or setting logic, but it might be simpler to just always set `*first_event=false` here?

Reply:
Good questions. From a quick search i cant find an exploration someone has done on that -- might be fun to test. In the meantime, I think you're right that it makes more sense to just set it.